### PR TITLE
Estimate Batchsize guide

### PR DESCRIPTION
**Pull request by 1297rohit**
\<insert pull request description here\>

-----

# Checklist

Please go through each entry in the below checklist and mark an 'X' if that condition has been met. Every entry should be marked with an 'X' to be get the Pull Request approved.

- [ ] All `import`s are in the first cell? 
    - [ ] First block of imports are standard libraries
    - [ ] Second block are 3rd party libraries
    - [ ] Third block are all `arcgis` imports? Note that in some cases, for samples, it is a good idea to keep the imports next to where they are used, particularly for uncommonly used features that we want to highlight.
- [ ] All `GIS` object instantiations are one of the following?
    - `gis = GIS()`
    - `gis = GIS('home')` or `gis = GIS('pro')`
    - `gis = GIS(profile="your_online_portal")`
    - `gis = GIS(profile="your_enterprise_portal")`
- [ ] If this notebook requires setup or teardown, did you add the appropriate code to `./misc/setup.py` and/or `./misc/teardown.py`?
- [ ] If this notebook references any portal items that need to be staged on AGOL/Python API playground, did you coordinate with a Python API team member to stage the item the correct way with the `api_data_owner` user?
- [ ] If the notebook requires working with local data (such as CSV, FGDB, SHP, Raster files), upload the files as items to the [Geosaurus Online Org](geosaurus.maps.arcgis.com/) using `api_data_owner` account and change the notebook to first download and unpack the files.
- [ ] Code simplified & split out across multiple cells, useful comments?
- [ ] Consistent voice/tense/narrative style? Thoroughly checked for typos?
- [ ] All images used like `` instead of ``? All map widgets contain a static image preview? (Call `mapview_inst.take_screenshot()` to do so)
- [ ] All file paths are constructed in an OS-agnostic fashion with `os.path.join()`? (Instead of `r"\foo\bar"`, `os.path.join(os.path.sep, "foo", "bar")`, etc.)
- [ ] Is your code formatted using [Jupyter Black](https://www.freecodecamp.org/news/auto-format-your-python-code-with-black/)? You can use Jupyter Black to format your code in the  notebook.
- [ ] **IF YOU WANT THIS SAMPLE TO BE DISPLAYED ON THE DEVELOPERS.ARCGIS.COM WEBSITE**, ping @ mohi9282 so he can add it to the list for the next deploy

### CODE REVIEW

File: guide/14-deep-learning/how_estimate_batch_size_works.ipynb [hunk 1]
```diff
@@ -0,0 +1,251 @@
+{
```
Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>This guide explains the steps for calculating the&nbsp;<strong>recommended_batchsize</strong>&nbsp;and&nbsp;<strong>max_batchsize</strong>&nbsp;of multiple network architectures supported by the&nbsp;<code>arcgis.learn</code>&nbsp;API. The&nbsp;<code>estimate_batch_size</code>&nbsp;function is designed to determine the optimal batch size for a deep learning model. Batch size is a critical hyperparameter in training deep learning models, and finding the right value can significantly impact the model's performance. The function takes in an&nbsp;<code>arcgis.learn</code>&nbsp;imagery model instance for which a batch size should be estimated and returns a batch size based on GPU capacity, size of the model, and the data used. By using the&nbsp;<code>estimate_batch_size</code>&nbsp;function, deep learning practitioners can prevent memory errors that can occur during training when the batch size is set too high. By finding the optimal batch size for the given model and system resources, the function can help ensure that the model fits within the available memory capacity and prevent errors that can interrupt the training process. This can save time and resources by avoiding the need to manually adjust the batch size or troubleshoot memory-related issues.</p><p>Currently only <code>arcgis.learn</code> imagery models are supported by the function.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>Unnecessary new line here? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>Next, we will see how the <code>estimate_batch_size</code> function works with <code>arcgis.learn</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>Prepare the data for the <code>SingleShotDetector</code> class using <code>prepare_data()</code> in <code>arcgis.learn</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>The <code>estimate_batch_size</code> class accepts the following paramters: </p><ul><li><code>model</code>&nbsp;(Required Parameter): Required&nbsp;<code>arcgis.learn</code>&nbsp;imagery model. The model instance for which batch size should be estimated. Not supported for text, tabular, timeseries, or tracking models like <code>FullyConnectedNetwork</code>, <code>MLModel</code>, <code>TimeSeriesModel</code>, <code>SiamMask</code>, <code>PSETAE</code> and <code>EfficientDet</code> models.</li><li><code>mode</code>&nbsp;(Optional Parameter): Optional string. Default train. The mode for which batch size is estimated. Supports&nbsp;<code>train</code>&nbsp;and&nbsp;<code>eval</code>&nbsp;modes for calculating batch size in training mode and evaluation mode respectively. Note: <code>max_batchsize</code> is capped at <code>1024</code> for train and eval mode and <code>recommended_batchsize</code> is capped at <code>64</code> for train mode.</li><li><br></li><li class="ql-indent-1"><code>train</code> : To to be used when the user wants to estimate batch size for training a model.</li><li class="ql-indent-1"><code>eval</code> : To be used when the user also wants to estimate batch size for inferencing on a trained model.</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>The output of this function will be the <code>recommended_batchsize</code> and <code>max_batchsize</code> for training a <code>SingleShotDetector</code> model.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>The output of this function will be the <code>recommended_batchsize</code> and <code>max_batchsize</code> for inferencing on a trained <code>SingleShotDetector</code> model.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (BP-Ent) [inline guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>It is possible to use the <code>estimate_batch_size</code> function to determine the <code>recommended_batchsize</code> during the data preparation stage while using the <code>prepare_data</code> function. The <code>prepare_data</code> function typically prepares the data for model training by performing tasks like data cleaning, preprocessing, and splitting the data into training, validation, and test sets. To use <code>estimate_batch_size</code> from the <code>prepare_data</code> function while creating the data object using <code>prepare_data</code>, we need to pass <code>batch_size</code> as <code>None</code>. This will call the <code>estimate_batch_size</code> function within the <code>prepare_data</code> function, passing in the relevant data and model parameters as arguments, and setting the <code>recommended_batchsize</code> as the <code>batch_size</code> for the data object. This helps ensure that the model is trained using the best possible batch size, improving its performance and reducing the likelihood of memory errors.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/1533/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Review by BP-Ent — CHANGES_REQUESTED** [review]
Suggested changes made on reviewnb

File: guide/14-deep-learning/how_estimate_batch_size_works.ipynb (hunk 1 above)
Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>Added</p>

[the remark this answers is not in the document]
Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>it will work with gpu. It tries to find the batch size that can fit gpu</p>

Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>done</p>

[the remark this answers is not in the document]
Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>we can by using the normal data.batch_size</p>

[the remark this answers is not in the document]
Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>fixed it.</p>

File: guide/14-deep-learning/how_estimate_batch_size_works.ipynb [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
```
[the remark this answers is not in the document]
Comment (1297rohit) [reply guide/14-deep-learning/how_estimate_batch_size_works.ipynb:1]: <p>done</p>

**Comment by priyankatuteja** [comment]
@1297rohit closing this PR, as we can add this as a sub section in existing guides
